File: JS/day_7/part2.js
import { readFile } from "node:fs/promises";

async function part2(path) {
  const file = await readFile(path, "utf8");
  const data = file.trim().split("\n");

  const cols = data[0].length;
  let count = 0;

  let beams = Array(cols).fill(0);
  beams[data[0].indexOf("S")] = 1;

  let next = Array(cols).fill(0);

  for (const line of data.slice(1)) {
    next.fill(0);

    beams.forEach((n, col) => {
      if (n === 0) return;

      if (line[col] === "^") {
        col - 1 >= 0 ? (next[col - 1] += n) : (count += n);
        col + 1 < cols ? (next[col + 1] += n) : (count += n);
      } else {
        next[col] += n;
      }
    });

    [beams, next] = [next, beams];
  }

  count += beams.reduce((a, b) => a + b, 0);

  console.log(count);
}

part2("input.txt");

File: GO/day_7/part2.go
package day7

import (
	"bytes"
	"fmt"

	"aoc2025/utils"
)

func Part2(f string) {
	data := utils.ReadByteLines(path[f])
	cols := len(data[0])

	count := 0

	beams := make([]int, cols)
	beams[bytes.Index(data[0], []byte("S"))] = 1

	next := make([]int, cols)

	for _, line := range data[1:] {
		for i := range next {
			next[i] = 0
		}

		for col, n := range beams {
			if n == 0 {
				continue
			}

			switch line[col] {
			case '^':
				next[col+1] += n
				next[col-1] += n
			default:

				next[col] += n
			}
		}

		beams, next = next, beams
	}

	for _, n := range beams {
		count += n
	}

	fmt.Printf("Day 7 | Part 2: %d\n", count)
}
